builtin/v13/migration/market.go

```diff
@@ -0,0 +1,536 @@
+	if err != nil {
+		return cid.Undef, cid.Undef, xerrors.Errorf("failed to load prevOutProviderSectors map: %w", err)
+	}
+
+	// in-memory maps with changesets to be applied to prevOutProviderSectors
+	providerSectorsMem := map[abi.ActorID]map[abi.SectorNumber][]abi.DealID{}        // added
+	providerSectorsMemRemoved := map[abi.ActorID]map[abi.SectorNumber][]abi.DealID{} // removed
+
+	addProviderSectorEntry := func(deal abi.DealID, newState *market13.DealState) (abi.SectorNumber, error) {
+		sid, ok := m.providerSectors.dealToSector[deal]
+		if !ok {
+			return 0, xerrors.Errorf("deal %d not found in providerSectors", deal) // todo is this normal and possible??
+		}
+
+		newState.SectorNumber = sid.Number
+		if _, ok := providerSectorsMem[sid.Miner]; !ok {
+			providerSectorsMem[sid.Miner] = make(map[abi.SectorNumber][]abi.DealID)
+		}
+		providerSectorsMem[sid.Miner][sid.Number] = append(providerSectorsMem[sid.Miner][sid.Number], deal)
+
+		return sid.Number, nil
+	}
+
+	removeProviderSectorEntry := func(deal abi.DealID, newState *market13.DealState) error {
+		sid, ok := m.providerSectors.removedDealToSector[deal]
+		if !ok {
+			if newState.SectorNumber == 0 {
+				// already removed, sector was likely terminated/slashed, but not
+				// yet picked up by the market
+				return nil
+			}
+
+			// this can happen if the deal was terminated between premigration and migration
+			// if it's not here it wasn't in premigration ProviderSectors either
+			fmt.Printf("notice: deal %d not found in removedDealToSector\n", deal)
+			newState.SectorNumber = 0
+			return nil
+		}
+
+		newState.SectorNumber = 0
+		if _, ok := providerSectorsMemRemoved[sid.Miner]; !ok {
+			providerSectorsMemRemoved[sid.Miner] = make(map[abi.SectorNumber][]abi.DealID)
+		}
+		providerSectorsMemRemoved[sid.Miner][sid.Number] = append(providerSectorsMemRemoved[sid.Miner][sid.Number], deal)
+
+		return nil
+	}
+
+	//fmt.Printf("state diffs: %d\n", len(diffs))
+	//fmt.Printf("dealToSector: %d\n", len(m.providerSectors.dealToSector))
+	//fmt.Printf("removedDealToSector: %d\n", len(m.providerSectors.removedDealToSector))
+
+	var oldState, prevOldState market12.DealState
+	var newState market13.DealState
+	for _, change := range diffs {
+		deal := abi.DealID(change.Key)
+
+		switch change.Type {
+		case amt.Add:
+			if err := oldState.UnmarshalCBOR(bytes.NewReader(change.After.Raw)); err != nil {
+				return cid.Cid{}, cid.Cid{}, xerrors.Errorf("failed to unmarshal old state: %w", err)
+			}
+
+			newState.SlashEpoch = oldState.SlashEpoch
+			newState.LastUpdatedEpoch = oldState.LastUpdatedEpoch
+			newState.SectorStartEpoch = oldState.SectorStartEpoch
+			newState.SectorNumber = 0 // terminated / not found (?)
+
+			if oldState.SlashEpoch == -1 {
+				si, err := addProviderSectorEntry(deal, &newState)
+				if err != nil {
+					return cid.Cid{}, cid.Cid{}, xerrors.Errorf("failed to add provider sector entry: %w", err)
+				}
+				newState.SectorNumber = si
```

Comment:
My Go-Fu may be lacking but doesn't it make the `newState.SectorNumber = sid.Number` in `addProviderSectorEntry` redundant?